### ISSUE TITLE
drag-to-pan usually gets stuck

### ISSUE THREAD

**Issue by 0ion9** [occivink/mpv-image-viewer#2]
I have drag-to-pan bound as suggested in input.conf:

`MBTN_RIGHT script-binding image-viewer/drag-to-pan`

Usually this binding does not work -- ie. rightclicking or rightdragging doesn't do anything visible. However, operating the mousewheel to zoom in/out and then quickly rightdragging allows drag-to-pan to work. After that one operation, I must again perform the mousewheel -> drag dance if I want to pan again.

My conclusion is that cursor-centric-zoom (or a sideeffect of it) sets some part of mpv state in a way that allows drag-to-pan to work, but this state 'times out' (maybe something that is automatically reset by MPV). Reading the code, the only idea that occurs to me is the table["event"] == "down" check.

'pan follows cursor', which makes the same check, doesn't appear to work correctly, but fails in a slightly different, more usable way (it works as a oneshot operation that, as soon as the click comes in, takes the relative mouse position within the window as the point within the image to pan to, eg. clicking 0,0 in the window pans to the top left of image and the operation immediately ends)

MPV event API has changed in some subtle way, perhaps?

EDIT: it's also worth noting that other mpv scripts using dragging work completely normally -- eg, mpv_crop_script. I haven't yet found another script that tries to bind dragging directly (rather than within a distinct mode), but if you know of one, I'm happy to test it.

EDIT2: You'd think that 'window-dragging' might impact this, but window-dragging=yes or no, neither have any effect on the problem.

EDIT3: That's enough debugging for a bit. My current best guess is that there is some other valid state for table["event"] that should be caught.

mpv version: 0.29.0-75-gda1073c247

**Comment by occivink** [comment]
I just tried it with `master` and it works as expected. You may be right with your guess that the 'complex' key binding is not fully reported, the manual mentions that it's a possibility (https://mpv.io/manual/master/#lua-scripting-complex). Which platform are you on?
If you could add a `print(utils.format_json(table))` to one of the `pan_follows_cursor_handler` or `drag_to_pan_handler` that'd be helpful. 

Actually judging by the symptoms it looks like it's the `mouse_move` keybinding that does not work, but it's used by the osc so I can't imagine that being not reliable. Can you try this short script and check that it works (prints the mouse position to the console)?
```
mp.add_forced_key_binding("mouse_move", "test", function()
print(mp.get_mouse_pos())
end)
```

**Comment by jgreco** [comment]
I don't know if it's related to 0ion9's troubles, but for me drag-to-pan typically only works if the initiating click is in the upper left quadrant of an image.  If the image is zoomed such that the lower right corner of the image is in the upper left of the window, then I can click there to drag and the drag will follow across the full extent of the screen, but initiating drags in the lower right of the window doesn't work.

I've been meaning to investigate this more, but haven't gotten around to it yet.   Maybe it's unrelated but if so it suggests to me there might be something quite screwy going on in mpv's handling of these sort of mouse events.

(this is on OSX for me)

**Comment by occivink** [comment]
Ok I was just testing around and it seems like the OSC is intercepting mouse clicks, such that the script never gets the events. @0ion9 can you confirm that you have the OSC enabled, and whether disabling it (`osc=no`) fixes it? 
I'm not sure if there's a better fix for it

**Comment by sammoth** [comment]
Not sure if this is related to the issues in this discussion, but when I click to drag or scroll to zoom using the features of this script, it will not work correctly unless the mouse has moved at least 1 pixel over the mpv window since mpv started. If it hasn't, the drag will move the image suddenly out of frame, and the zoom will always be centred on the top left corner of the image. I assume that when the mouse hasn't been moved, mpv doesn't pick up the actual position of the mouse on startup?

**Comment by occivink** [comment]
Yeah, I wasn't aware of this but it looks like `mp.get_mouse_pos()` reports a bogus `0, 0` if you haven't moved the mouse yet. A solution to that would be for mpv to initialize the mouse position  to `INT_MIN, INT_MIN` or something. 
If it really matters to you you can report a bug to mpv or even submit a PR.

**Comment by haebawari** [comment]
not sure if it's related to current issue, but for me right click drag-to-pan does not work in any region of the player where moving mouse brings up the bottom panel with seek-bar and playback controls.
it does work however in the top 2/5 parts of the screen where moving mouse does not trigger the bottom panel.

**Comment by dumblob** [comment]
I am now testing mouse pan with both `osc=no` and `osd-bar=no` and it works perfectly in a fullscreen view (`f` by default in mpv). But once I switch from fullscreen to windowed (again by pressing `f`) the panning by mouse does not work any more and instead it moves the whole window without changing its content.

Anybody any ideas how to fix this?

**Comment by butteria** [comment]
@dumblob the whole window? to me, right-click doesn't work just like what metioned above, you can try to bind the function with 'ctrl+MBTN_RIGHT' instead of 'MBTN_RIGHT', or add this code 'mp.commandv("script-message", "osc-visibility" ,"never")' before  'mp.add_forced_key_binding("mouse_move", "image-viewer-mouse-move", function() moved = true end)'. i fix it but i don't know the reason. maybe it's the confliction betwenn bulit-in mouse button and osc. because i use 'h' to trigger the funtion and it works too.

**Comment by dumblob** [comment]
@butteria I forgot to mention I have tested panning with `MBTN_LEFT  script-binding drag-to-pan` in `input.conf`.

Now I have tried it with `MBTN_RIGHT  script-binding drag-to-pan` and **it works** as expected even in a non-fullscreen window.

Therefore I would call this (and similar) a special bug as `mpv` seems to harcode some behavior of certain keys (be it on mouse or keyboard or combination thereof).

Any ideas how to circumvent this "buggy" behavior? I would appreciate even very hacky solutions or workarounds :wink:.

**Comment by dumblob** [comment]
Note that `mp.commandv("script-message", "osc-visibility" ,"never")` trick should not apply to my case as I have disabled `osc` altogether by `osc=no` in `mpv.conf`.

That said I have tried this trick and it does not change any of the behavior I was observing without this trick. Thanks anyway @butteria .

Any other ideas?
